aio: add renderExamples processor

petebacondarwin opened the pull request:
build(aio): add renderExamples processor  …
This processor will eventually replace the `{@example}` inline tags
because it provides a cleaner approach that also supports tabbed examples
straight out of the box.

The idea is that authors will simply add a `path` and (optionally) a `region`
attribute to `<code-example>` or `<code-pane>` elements in their docs.
This indicates to dgeni that the relevant example needs to be injected
into the content of this element.

For example, assume that there is an example file `toh-pt1/index.hml` with
a region called `title`, which looks like:

```
<!-- #docregion title -->
<h1>Tour of Heroes</h1>
<!-- #enddocregion title -->
```

Then the document author could get this to appear in the docs as a
standalone example:

```
<code-example path="toh-pt1" region="title"></code-example>
```

Or as part of a tabbed group:

```
<code-tabs>
  <code-pane path="toh-pt1" region="title"></code-pane>
</code-tabs>
```

If no `path` attribute is provided then the element content is ignored, which
enables authors to provide inline code instead:

```
<code-example>
 Some &lt;html&gt escaped code
</code-example>
```

Also all attributes other than `path` and `region` are ignored and passed
through to the final rendered output allowing the author to provide
styling hints:

```
<code-example path="toh-pt1" region="title" linenums="15" class="important">
</code-example>
```

petebacondarwin committed:
build(aio): refactor getExampleRegion into a reusable service

mary-poppins commented:
The angular.io preview for 39c7b55 is available [here][1].

[1]: https://pr15425-39c7b555dac80a580bf6e753451eb6908033e604.ngbuilds.io/

gkalpak commented on aio/transforms/examples-package/processors/render-examples.js:
```diff
@@ -0,0 +1,50 @@
+
+function parseAttributes(str) {
+  const attrMap = {};
+  str.split(' ')
```
What about spaces inside the attribute value (e.g. `class="foo bar"`)? Do we care to support that?

wardbell replied:
Yes ... we must support that.

wardbell replied:
Yes. We will have `class="foo bar"`. We will have `title="This is a title"`.

wardbell replied:
Maybe add this test:
```
  it('should remove the path and region attributes but leave the other attributes - with spaces - alone', () => {
    const docs = [
      { renderedContent: '<code-example class="special" path="test/url" linenums="15" region="region-1" class="foo bar" title="The Best">Some code</code-example>' }
    ];
    processor.$process(docs);
    expect(docs[0].renderedContent).toEqual('<code-example class="special" linenums="15" class="foo bar" title="The Best">\nregion 1 contents\n</code-example>');
  });
```

gkalpak commented on aio/transforms/examples-package/processors/render-examples.js:
```diff
@@ -0,0 +1,50 @@
+              const exampleContent = getExampleRegion(doc, attrMap.path, attrMap.region);
+              delete attrMap.path;
+              delete attrMap.region;
+              attributes = Object.keys(attrMap).map(key => ` ${key}="${attrMap[key]}"`).join('');
```
Too far-fetched, but this will break for attributes that contain `"` in their value and use a different delimiter (e.g. `foo='double"quotes'`).

petebacondarwin replied:
Yep. Good spot

wardbell replied:
Not so far fetched. I think it is possible that we'll see quoted stuff like this
```
<code-pane path="test/bob.ts" region="foo" title="bob.ts ('foo')"><code-pane>
```
Because we can't put html within attributes, I don't see how it can be escaped.

If it helps, we could require the author to follow some kind of escape scheme that is easier to parse such as backslashing:
```
<code-pane path="test/bob.ts" region="foo" title="bob.ts (\'foo\')"><code-pane>
<code-example path="test/bob.ts" region="slash" title="bob.ts - using backslashes ([\\]) for fun"><code-example>
```

gkalpak commented on aio/transforms/examples-package/processors/render-examples.spec.js:
```diff
@@ -0,0 +1,73 @@
+    expect(processor.$runAfter).toEqual(['docs-rendered']);
+  });
+
+  it('should ignore a <code-example> tags with no path attribute', () => {
```
_a ... tags_ --> _... tags_

gkalpak replied:
And no tests for `<code-pane>`. I can't even imagine how `<code-pane>`s must be feeling right now 😞 
That kind of discrimination is **not** OK!

gkalpak reviewed (commented):
Nice 💯

wardbell commented on aio/transforms/examples-package/services/getExampleRegion.js:
```diff
@@ -0,0 +1,31 @@
+
+    // Find the example in the folders
+    var exampleFile;
+    // Try an "annotated" version first
```
What is an "annotated version"?

petebacondarwin replied:
Do you remember we had long discussions about adding regions to JSON files?
I didn't like the idea of using regexes to access the JSON, so I came up with the idea of annotated files, which allows you to create an annotated file next to the real file that contains the doc region markers.
You didn't like that but we settled on the fact that we also pull in the whole file if there are no region markers and that was enough for the current author requirements.
But I left the annotated files stuff in there anyway, just in case.

So to be clear, if you have a file that cannot contain comments (e.g. `package.json`) then you can create a copy with annotations (e.g. `package.json.annotated`) which does contain comments with doc region markers.

If the example collector finds such an annotated file it will use that rather than the original.

wardbell reviewed (changes requested):
Looking good. We just need to address the concern about spaces and quotes mentioned in the feedback.

mary-poppins commented:
The angular.io preview for f20ff13 is available [here][1].

[1]: https://pr15425-f20ff13d2d9dfe99e3730773b423c92014979204.ngbuilds.io/

gkalpak commented on aio/transforms/examples-package/utils.js:
```diff
@@ -3,5 +3,83 @@ module.exports = {
+    function takeAttribute() {
+      if (index >= 0 && index < str.length) {
+        takeKey();
+        skipSpace();
```
Theoretically, no space is allowed between the key and the `=`.

petebacondarwin replied:
I don't believe that is true: https://www.w3.org/TR/html51/syntax.html#elements-attributes
And reading through that spec I realise that I have not accounted for the "empty attribute syntax" nor the "unquoted attribute value syntax"

gkalpak replied:
Wow, I thought browsers choke on whitespace around the equals sign, but apparently they don't :grin:

gkalpak commented on aio/transforms/examples-package/utils.js:
```diff
@@ -3,5 +3,83 @@ module.exports = {
+
+    function expectQuote() {
+      quote = str[index];
+      if (!['"', "'"].indexOf(quote) === -1) {
```
`!... === -1` --> `... !== -1`

petebacondarwin replied:
Actually it was completely wrong. It should be:

```
if (['"', "'"].indexOf(quote) === -1) {
```

gkalpak commented on aio/transforms/examples-package/utils.js:
```diff
@@ -3,5 +3,83 @@ module.exports = {
+    const attrMap = {};
+    let index = 0, key, quote, value;
+
+    while(0 <= index && index < str.length) {
```
Nit: Since `index` is initialized at 0 and only incremented, the >=0 check is not necessary (here and below).

petebacondarwin replied:
Originally I had some lines that looked like: `index = str.search(someChar, index)`

gkalpak commented on aio/transforms/examples-package/utils.js:
```diff
@@ -3,5 +3,83 @@ module.exports = {
     return mappedObj;
+  },
+
+  parseAttributes(str) {
```
Oh, a tiny parser :heart:

petebacondarwin replied:
It is complex enough not to do easily with straight regex.
I am sure that a good hacker could improve its speed considerably.

petebacondarwin commented on aio/transforms/examples-package/inline-tag-defs/example.js:
```diff
@@ -14,16 +14,14 @@ var entities = require('entities');
  */
 module.exports = function exampleInlineTagDef(
-    parseArgString, exampleMap, createDocMessage, log, collectExamples) {
+    parseArgString, exampleMap, createDocMessage, log, collectExamples, getExampleRegion) {
```
no - good spot

petebacondarwin replied:
and neither is `log` nor `exampleMap` :-)

mary-poppins commented:
The angular.io preview for 797523b is available [here][1].

[1]: https://pr15425-797523bd494e043f9a0dbbd971b4cb57f7332eb2.ngbuilds.io/

gkalpak commented on aio/transforms/examples-package/utils.js:
```diff
@@ -1,31 +1,43 @@
+      } else {
+        attrMap[key] = true;
       }
+      // skip the closing quote or whitespace
```
This comment seems misplaced

petebacondarwin replied:
Probably. I'll look into it

gkalpak commented on aio/transforms/examples-package/utils.js:
```diff
@@ -35,51 +47,47 @@ module.exports = {
+          index++;
+        }
+        if (index >= str.length) {
+          throw new Error(`Unterminated quoted attribute value in \`${str}\`. Starting at ${startIndex}. Expected a ${quote} but got "end of string"`);
```
Missing punctuation at the end 😱

mary-poppins commented:
The angular.io preview for 2916b02 is available [here][1].

[1]: https://pr15425-2916b0249ee0b1b5fa5e6e9e19ac412e8a373abb.ngbuilds.io/

petebacondarwin committed:
build(aio): add renderExamples processor

This processor will eventually replace the `{@example}` inline tags
because it provides a cleaner approach that also supports tabbed examples
straight out of the box.

The idea is that authors will simply add a `path` and (optionally) a `region`
attribute to `<code-example>` or `<code-pane>` elements in their docs.
This indicates to dgeni that the relevant example needs to be injected
into the content of this element.

For example, assume that there is an example file `toh-pt1/index.hml` with
a region called `title`, which looks like:

```
<h1>Tour of Heroes</h1>
```

Then the document author could get this to appear in the docs as a
standalone example:

```
<code-example path="toh-pt1" region="title"></code-example>
```

Or as part of a tabbed group:

```
<code-tabs>
  <code-pane path="toh-pt1" region="title"></code-pane>
</code-tabs>
```

If no `path` attribute is provided then the element is ignored, which
enables authors to provide inline code instead:

```
<code-example>
 Some &lt;html&gt escaped code
</code-example>
```

Also all attributes other than `path` and `region` are ignored and passed
through to the final rendered output allowing the author to provide
styling hints:

```
<code-example path="toh-pt1" region="title" linenums"15" class="important">
</code-example>
```

mary-poppins commented:
The angular.io preview for 963e2fe is available [here][1].

[1]: https://pr15425-963e2fe7a683c5a4b64134939421cbe556bce303.ngbuilds.io/
